### PR TITLE
Create bottled cards command

### PR DESCRIPTION
**Pull request by brianyee**
Create the !bottled command, this should list the bottled cards for the base game Bottle relics. 
I believe there may be modded Bottle relics but this does not cover them.

Bottled cards also have the "misc" property for their current value if it's a meta-scaling card (Ritual Dagger, Genetic Algorithm, etc), but I don't think the actual value needs to be part of the command

### CODE REVIEW

File: save.py [hunk 1]
```diff
@@ -195,6 +196,24 @@ def rare_chance_as_str(self) -> tuple[str, str, str]:
     def upcoming_boss(self) -> str:
         return self._data["boss"]
 
+    @property
+    def bottled_cards(self) -> str:
+        cards = []
+        if self._data["bottled_flame"]:
```
Comment (brianyee) [inline save.py:202]: I don't actually know if this checks existence of the json key. The "bottled_x" keys only exist once you pick up the relic.

File: server.py [hunk 2]
```diff
@@ -832,6 +832,12 @@ async def skipped_boss_relics(ctx: ContextType, save: Savefile): # JSON_FP_PROP
 
     await ctx.send(" ".join(msg))
 
+@with_savefile("bottled")
+async def bottled_cards(ctx: ContextType, save: SaveFile):
+    """List all bottled cards"""
+    cards = save.bottled_cards
+    await ctx.send(", ".join(cards) if len(cards) > 0 else "We do not have any bottled cards.")
```
Comment (brianyee) [inline server.py:839]: Feel free to change the verbiage or logic however you think is better pattern-wise or anything

File: save.py [hunk 3]
```diff
@@ -17,6 +17,7 @@
 from logger import logger
 from utils import get_req_data
 from runs import get_latest_run
+from nameinternal import get_card
```
Comment (Vgr255) [inline save.py:20]: Nitpick: imports should be ordered in descending order of module name length

**Comment by brianyee** [comment]
I feel the need to note that this is completely untested lol

File: save.py [hunk 4]
```diff
@@ -195,6 +196,24 @@ def rare_chance_as_str(self) -> tuple[str, str, str]:
...
+        cards = []
+        if self._data["bottled_flame"]:
+            cards.append(self.__get_card_string(self._data["bottled_flame"], self._data["bottled_flame_upgrade"]))
+        if self._data["bottled_lightning"]:
+            cards.append(self.__get_card_string(self._data["bottled_lightning"], self._data["bottled_lightning_upgrade"]))
+        if self._data["bottled_tornado"]:
+            cards.append(self.__get_card_string(self._data["bottled_tornado"], self._data["bottled_tornado_upgrade"]))
+        return cards    
+
+    def __get_card_string(self, card: str, upgrades: int) -> str:
```
Comment (Vgr255) [inline save.py:210]: Please no double leading underscore!

File: save.py [hunk 5]
```diff
@@ -195,6 +196,24 @@ def rare_chance_as_str(self) -> tuple[str, str, str]:
...
+        if upgrades == 1:
+            card += "+"
+        elif upgrades > 1:
+            card += "+" + str(upgrades)
```
Comment (Vgr255) [inline save.py:214]: `get_card` automatically converts `card+1` into `card+`. The following would probably work better:
```
    if upgrades:
        card = f"{card}+{upgrades}"
    return get_card(card)
```

File: server.py [hunk 6]
```diff
@@ -832,6 +832,12 @@ async def skipped_boss_relics(ctx: ContextType, save: Savefile): # JSON_FP_PROP
 
     await ctx.send(" ".join(msg))
 
+@with_savefile("bottled")
+async def bottled_cards(ctx: ContextType, save: SaveFile):
+    """List all bottled cards"""
```
Comment (Vgr255) [inline server.py:837]: Docstrings should end in a period.

File: server.py (hunk 2 above)
Comment (Vgr255) [inline server.py:839]: I'd prefer the `not cards` part to be a different line, rather than this clever one-liner (as much as I love clever one-liners)

**Review by Vgr255 — CHANGES_REQUESTED** [review]

File: save.py (hunk 1 above)
Comment (Vgr255) [reply save.py:202]: It does not. You want `if self._data("bottled_flame")`

File: save.py (hunk 4 above)
Comment (brianyee) [reply save.py:210]: do we have a "private" method pattern? I just read this online. I didn't feel the need for this to be exposed anywhere else

Comment (Vgr255) [reply save.py:210]: Yeah, just single leading underscore is fine. Double-leading underscore does some magic with the name, which I don't want.

**Comment by brianyee** [comment]
> How about adding which bottle the cards are from for UX clarity? I personally like using emoji:
> 
> ⚡ Apotheosis, 🔥 Searing Blow +69, 🌪️ Corruption
> 
> All of those are standard emoji, so they should work "everywhere". (baller run if we have all of the above btw)

I have never used emojis directly in code, is it fine to just have it like this then?
image

Or is there some like ":tornado:" colon code I should use? I like the idea of specifying them though

File: save.py [hunk 7]
```diff
@@ -195,6 +196,22 @@ def rare_chance_as_str(self) -> tuple[str, str, str]:
     def upcoming_boss(self) -> str:
         return self._data["boss"]
 
+    @property
+    def bottled_cards(self) -> str:
+        cards = []
+        if self._data("bottled_flame"):
+            cards.append("🔥 " + self._get_card_string(self._data["bottled_flame"], self._data["bottled_flame_upgrade"]))
```
Comment (Vgr255) [inline save.py:203]: I would rather use the emoji names than putting the actual emoji in (mostly as a matter of personal preference). The way to do this is the following (capital letters necessary):

🔥 = `"\N{FIRE}"` 
⚡ = `"\N{HIGH VOLTAGE SIGN}"`
🌪️ = `"\N{CLOUD WITH TORNADO}"`

File: server.py [hunk 8]
```diff
@@ -832,6 +832,15 @@ async def skipped_boss_relics(ctx: ContextType, save: Savefile): # JSON_FP_PROP
 
     await ctx.send(" ".join(msg))
 
+@with_savefile("bottle", "bottled")
+async def bottled_cards(ctx: ContextType, save: SaveFile):
+    """List all bottled cards."""
+    cards = save.bottled_cards
+    if len(cards) > 0:
```
Comment (Vgr255) [inline server.py:839]: Using `len` to check for emptiness is bad practice. Just say `if cards` instead!

**Review by Vgr255 — APPROVED** [review]

**Pull request merged**